stl/inc/expected

```diff
@@ -685,13 +700,393 @@ public:
+        static_assert(
+            is_convertible_v<_Uty, _Err>, "is_convertible_v<G, E> must be true. (N4928 [expected.object.obs]/22)");
+
+        if (_Has_value) {
+            return _STD forward<_Uty>(_Other);
+        } else {
+            return _STD move(_Unexpected);
+        }
+    }
+
+    // [expected.object.monadic]
+    template <class _Fn>
+        requires is_copy_constructible_v<_Err>
+    constexpr auto and_then(_Fn&& _Func) & {
+        using _Uty = remove_cvref_t<invoke_result_t<_Fn, _Ty&>>;
+
+        static_assert(_Is_specialization_v<_Uty, expected>,
+            "expected<T,E>::and_then(F) requires the return type of F to be a specialization of expected. "
+            "(N4928 [expected.object.monadic]/3).");
+        static_assert(is_same_v<typename _Uty::error_type, _Err>,
+            "expected<T,E>::and_then(F) requires the error type of the return type of F to be E. "
+            "(N4928 [expected.object.monadic]/3)");
+
+        if (_Has_value) {
+            return _STD invoke(_STD forward<_Fn>(_Func), _Value);
+        } else {
+            return _Uty(unexpect, _Unexpected);
+        }
+    }
+
+    template <class _Fn>
+        requires is_copy_constructible_v<_Err>
+    constexpr auto and_then(_Fn&& _Func) const& {
+        using _Uty = remove_cvref_t<invoke_result_t<_Fn, const _Ty&>>;
+
+        static_assert(_Is_specialization_v<_Uty, expected>,
+            "expected<T,E>::and_then(F) requires the return type of F to be a specialization of expected. "
+            "(N4928 [expected.object.monadic]/3).");
+        static_assert(is_same_v<typename _Uty::error_type, _Err>,
+            "expected<T,E>::and_then(F) requires the error type of the return type of F to be E. "
+            "(N4928 [expected.object.monadic]/3)");
+
+        if (_Has_value) {
+            return _STD invoke(_STD forward<_Fn>(_Func), _Value);
+        } else {
+            return _Uty(unexpect, _Unexpected);
+        }
+    }
+
+    template <class _Fn>
+        requires is_move_constructible_v<_Err>
+    constexpr auto and_then(_Fn&& _Func) && {
+        using _Uty = remove_cvref_t<invoke_result_t<_Fn, _Ty>>;
+
+        static_assert(_Is_specialization_v<_Uty, expected>,
+            "expected<T,E>::and_then(F) requires the return type of F to be a specialization of expected. "
+            "(N4928 [expected.object.monadic]/7).");
+        static_assert(is_same_v<typename _Uty::error_type, _Err>,
+            "expected<T,E>::and_then(F) requires the error type of the return type of F to be E. "
+            "(N4928 [expected.object.monadic]/7)");
+
+        if (_Has_value) {
+            return _STD invoke(_STD forward<_Fn>(_Func), _STD move(_Value));
+        } else {
+            return _Uty(unexpect, _STD move(_Unexpected));
+        }
+    }
+
+    template <class _Fn>
+        requires is_move_constructible_v<_Err>

```

Comment:
Maybe LWG issue: I feel like the constraint here should be `is_copy_constructible<_Err>` (or `is_move_constructible<const _Err>`) (same for `or_else`, `transform`, etc.)

Reply:
Yes, these constraints are bogus. I'd `is_move_constructible_v<E, decltype((unex))>` for the constraint in the issue resolution so it remains correct for both `&&` and `const&&` cases.

Reply:
So what's the protocol here? Do we fix our code to be "correct" and file an DR? Or file a DR and await resolution before fixing.

Reply:
We talked about this in the video code review, and while this is definitely an issue, it is surprisingly subtle, so we should file an LWG issue and wait for it to be resolved. It could be reasonable to drop the Constraints on these overloads entirely (what overload resolution scenario would they be improving?).

This also applies to `or_else`, `transform`, and `transform_error`.

Reply:
This is LWG-3877.